parallel-checkout.c

```diff
@@ -0,0 +1,1363 @@
+#include "cache.h"
+#include "blob.h"
+#include "object-store.h"
+#include "dir.h"
+#include "streaming.h"
+#include "submodule.h"
+#include "progress.h"
+#include "fsmonitor.h"
+#include "argv-array.h"
+#include "pkt-line.h"
+#include "quote.h"
+#include "sigchain.h"
+#include "sub-process.h"
+#include "trace2.h"
+#include "parallel-checkout.h"
+#include "checkout--helper.h"
+#include "thread-utils.h"
+#include "unpack-trees.h"
+
+int is_eligible_for_parallel_checkout(const struct conv_attrs *ca)
+{
+	enum conv_attrs_classification c = classify_conv_attrs(ca);
+
+	switch (c) {
+	default:
+		BUG("unsupported conv_attrs classification '%d'", c);
+		return 0;
+
+	case CA_CLASS_INCORE:
+		return 1;
+
+	case CA_CLASS_INCORE_FILTER:
+		/*
+		 * It would be safe to allow concurrent instances of
+		 * single-file smudge filters, like rot13, but we should
+		 * not assume that all filters are parallel-process safe.
+		 * So we don't allow this.
+		 */
+		return 0;
+
+	case CA_CLASS_INCORE_PROCESS:
+		/*
+		 * The parallel queue and the delayed queue are not
+		 * compatible and must be keep completely separate.
+		 * So "process" filtered items are not eligible for
+		 * parallel checkout.
+		 *
+		 * The process filter may return the blob content on the
+		 * initial request or it may return a "delayed" response
+		 * and not return the blob content until the CE_RETRY
+		 * request.  And we can't tell which without actually
+		 * asking the long-running process filter to do it.
+		 *
+		 * Furthermore, there should only be one instance of
+		 * the long-running process filter because we don't know
+		 * how it is managing its own concurrency.  And if we
+		 * spread the parallel queue across multiple threads or
+		 * processes, each would need to talk to their own
+		 * instance of the process filter and that could cause
+		 * problems.
+		 */
+		return 0;
+
+	case CA_CLASS_STREAMABLE:
+		return 1;
+	}
+}
+
+enum helper_spread_model {
```

Comment:
The sync mode would not want to modify it -- the whole point is that while the fg process is working on item[k] that the helpers are concurrently preloading blobs [k+1, k + 2, ... k + h-1] so that they are ready when the fg process advances to item[k+1].

But yeah, for Async Mode, it would be nice to tweak this (and I have done a few interactive experiments in that spirit, but it didn't make much difference),  But yeah, it would be nice to have this.  It's one of those things that might be filesystem dependent (as in, how does the kernel/fs do locking WRT directory inodes).